Add CI for the webhook

wainersm opened the pull request:
The webhook is a component that can be tested in isolation. You will only required a working Kubernetes cluster (which can be setup with Kind) and the tools to build the webhook itself. So this pull request contain an CI for the webhook which is triggered only and only if the webhook files changed.

The CI is built on a github action, runner script and bats tests. Developers will be able to run the same workflow locally by issuing `make test-e2e`.

Refer to the https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/3268928340 for a successful execution of the github action.

Note: there is a test case not passing which was marked to skip.

Fixes #291

bpradipt commented on webhook/Makefile:
```diff
@@ -94,12 +102,30 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
+	kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v1.6.1/cert-manager.yaml
+	@echo Connect the registry in the cluster network
+	docker network connect "kind" "${KIND_REG_NAME}"
+
```
I don't think a local registry is needed. The `kind-load` target can be used to load the newly built image and run the tests.

wainersm replied:
I tried to use `kind-load` with an image named `localhost/peer-pods-webhook:test` that didn't work. K8s was trying to pull the image with https, like `https://localhost` and this is not resolving. Is there anything that I am missing?

bpradipt replied:
Does this help - https://kind.sigs.k8s.io/docs/user/quick-start/#loading-an-image-into-your-cluster ?

wainersm replied:
Hi @bpradipt, please help me understand what I am doing wrong :)

I'm running the following commands:

```
export IMG=localhost/peerpods-webhook-test
make docker-build
make kind-cluster
make kind-load
make deploy
```
Just to be sure, I am pasting here the output of `kind-load`
```
$ make kind-load
kind load docker-image localhost/peerpods-webhook-test --name webhook
Image: "localhost/peerpods-webhook-test" with ID "sha256:59959471611ed7183b947197b6057e469a1367a463925e135f70eddd686a99b2" not yet present on node "webhook-control-plane", loading...
```

Then I get the following error; the image is not pulled:

```
$ kubectl describe pod/peer-pods-webhook-controller-manager-678fc8f46b-7zc6j -n peer-pods-webhook-systemName:         peer-pods-webhook-controller-manager-678fc8f46b-7zc6j
Namespace:    peer-pods-webhook-system
Priority:     0
Node:         webhook-control-plane/172.18.0.2
Start Time:   Tue, 25 Oct 2022 12:22:59 -0300
Labels:       control-plane=controller-manager
              pod-template-hash=678fc8f46b
Annotations:  kubectl.kubernetes.io/default-container: manager
Status:       Pending
IP:           10.244.0.8
IPs:
  IP:           10.244.0.8
Controlled By:  ReplicaSet/peer-pods-webhook-controller-manager-678fc8f46b
Containers:
  kube-rbac-proxy:
    Container ID:  containerd://cb0844278b41d8f6e2e5119cab7daf12fd9404fae57a6df9c9665d2cb70f1c17
    Image:         gcr.io/kubebuilder/kube-rbac-proxy:v0.11.0
    Image ID:      gcr.io/kubebuilder/kube-rbac-proxy@sha256:0df4ae70e3bd0feffcec8f5cdb428f4abe666b667af991269ec5cb0bbda65869
    Port:          8443/TCP
    Host Port:     0/TCP
    Args:
      --secure-listen-address=0.0.0.0:8443
      --upstream=http://127.0.0.1:8080/
      --logtostderr=true
      --v=0
    State:          Running
      Started:      Tue, 25 Oct 2022 12:23:11 -0300
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  128Mi
    Requests:
      cpu:        5m
      memory:     64Mi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jkj99 (ro)
  manager:
    Container ID:  
    Image:         localhost/peerpods-webhook-test:latest
    Image ID:      
    Port:          9443/TCP
    Host Port:     0/TCP
    Command:
      /manager
    Args:
      --health-probe-bind-address=:8081
      --metrics-bind-address=127.0.0.1:8080
      --leader-elect
    State:          Waiting
      Reason:       ImagePullBackOff
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  128Mi
    Requests:
      cpu:      10m
      memory:   64Mi
    Liveness:   http-get http://:8081/healthz delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:  http-get http://:8081/readyz delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment:
      TARGET_RUNTIMECLASS:       kata-remote-cc
      POD_VM_INSTANCE_TYPE:      t2.small
      POD_VM_EXTENDED_RESOURCE:  kata.peerpods.io/vm
    Mounts:
      /tmp/k8s-webhook-server/serving-certs from cert (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jkj99 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  webhook-server-cert
    Optional:    false
  kube-api-access-jkj99:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                   From               Message
  ----     ------       ----                  ----               -------
  Normal   Scheduled    5m17s                 default-scheduler  Successfully assigned peer-pods-webhook-system/peer-pods-webhook-controller-manager-678fc8f46b-7zc6j to webhook-control-plane
  Warning  FailedMount  5m17s                 kubelet            MountVolume.SetUp failed for volume "cert" : secret "webhook-server-cert" not found
  Normal   Pulling      5m16s                 kubelet            Pulling image "gcr.io/kubebuilder/kube-rbac-proxy:v0.11.0"
  Normal   Pulled       5m5s                  kubelet            Successfully pulled image "gcr.io/kubebuilder/kube-rbac-proxy:v0.11.0" in 10.717261872s
  Normal   Created      5m5s                  kubelet            Created container kube-rbac-proxy
  Normal   Started      5m5s                  kubelet            Started container kube-rbac-proxy
  Warning  Failed       4m20s (x3 over 5m5s)  kubelet            Failed to pull image "localhost/peerpods-webhook-test:latest": rpc error: code = Unknown desc = failed to pull and unpack image "localhost/peerpods-webhook-test:latest": failed to resolve reference "localhost/peerpods-webhook-test:latest": failed to do request: Head "http://localhost/v2/peerpods-webhook-test/manifests/latest": dial tcp [::1]:80: connect: connection refused
  Warning  Failed       4m20s (x3 over 5m5s)  kubelet            Error: ErrImagePull
  Warning  Failed       3m50s (x5 over 5m4s)  kubelet            Error: ImagePullBackOff
  Normal   Pulling      3m37s (x4 over 5m5s)  kubelet            Pulling image "localhost/peerpods-webhook-test:latest"
  Normal   BackOff      12s (x20 over 5m4s)   kubelet            Back-off pulling image "localhost/peerpods-webhook-test:latest"
```

Then I named the image without `localhost` prefix and ran again. The new error is related with the fact that it tries to pull the image from docker.io:
```
  Warning  Failed       4s (x2 over 18s)  kubelet            Failed to pull image "peerpods-webhook-test:latest": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/library/peerpods-webhook-test:latest": failed to resolve reference "docker.io/library/peerpods-webhook-test:latest": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
  Warning  Failed       4s (x2 over 18s)  kubelet            Error: ErrImagePull

```

bpradipt replied:
The reasoning is described here https://kind.sigs.k8s.io/docs/user/quick-start/#loading-an-image-into-your-cluster
and I'm reproducing the text here for ease of reference:

```NOTE: The Kubernetes default pull policy is IfNotPresent unless the image tag is :latest or omitted (and implicitly :latest) in which case the default policy is Always. IfNotPresent causes the Kubelet to skip pulling an image if it already exists. If you want those images loaded into node to work as expected, please:

don't use a :latest tag
and / or:

specify imagePullPolicy: IfNotPresent or imagePullPolicy: Never on your container(s).
```

We use `latest` tag and there is no `magePullPolicy` specified in the deployment manifest `webhook/config/manager/manager.yaml` and this results in the image being pulled always.

You'll need to patch to the deployment manifest like shown below and able to proceed.

```
diff --git a/webhook/config/manager/manager.yaml b/webhook/config/manager/manager.yaml
index 95bf688..db6c2a3 100644
--- a/webhook/config/manager/manager.yaml
+++ b/webhook/config/manager/manager.yaml
@@ -32,6 +32,7 @@ spec:
         args:
         - --leader-elect
         image: controller:latest
+        imagePullPolicy: IfNotPresent
         name: manager
         env:
         - name: TARGET_RUNTIMECLASS
```

Hope this helps..

wainersm replied:
Thanks for this kind RTFM @bpradipt ! :)
I changed the name of the image and it worked as expected without a registry. \o/

bpradipt commented on webhook/Makefile:
```diff
@@ -94,12 +102,30 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
+	@echo Connect the registry in the cluster network
+	docker network connect "kind" "${KIND_REG_NAME}"
+
+.PHONY: kind-cluster-delete
```
how about just naming it to `kind-delete` ?

wainersm replied:
works for me: shorter name, better name.

wainersm commented:
Updated this to remove the handling of an unnecessary local registry. Thanks @bpradipt for the review!
Added a new commit (https://github.com/confidential-containers/cloud-api-adaptor/pull/310/commits/38f137b511146d6d2e2570ab219b0572a541f812) to pin the k8s version used on the e2e tests.

bpradipt commented on webhook/docs/DEVELOPMENT.md:
````diff
@@ -87,3 +89,10 @@ and peer pod vm limit added.
+and then run:
+```sh
+make test-e2e
+```
````
nice

bpradipt reviewed (approved):
/lgtm
Thanks @wainersm

huoqifeng reviewed (approved):
/lgtm
thank you @wainersm and @bpradipt , I learnt some kind tricks while review this PR :-).
